libraries/AP_TECS/AP_TECS.cpp

```diff
@@ -1168,22 +1163,28 @@ void AP_TECS::_initialise_states(int32_t ptchMinCO_cd, float hgt_afe)
         _hgt_dem              = hgt_afe;
         _hgt_dem_in_prev      = hgt_afe;
         _hgt_dem_in_raw       = hgt_afe;
-        _TAS_dem_adj          = _TAS_dem;
-        _flags.reset          = true;
-        _post_TO_hgt_offset   = _climb_rate * _hgt_dem_tconst;
         _flags.underspeed     = false;
         _flags.badDescent     = false;
 
-        _max_climb_scaler = 1.0f;
-        _max_sink_scaler = 1.0f;
+        if (_in_takeoff_first_stage()) {
+            _post_TO_hgt_offset   = _climb_rate_limit * _hgt_dem_tconst; // Replacement prevents oscillating hgt_rate_dem.
+            _TAS_dem_adj = _TAS_dem;
+            _max_climb_scaler = 1.0f;
+            _max_sink_scaler = 1.0f;
+            _pitch_demand_lpf.reset(_ahrs.get_pitch());
+            _pitch_measured_lpf.reset(_ahrs.get_pitch());
+        }
 
-        _pitch_demand_lpf.reset(_ahrs.get_pitch());
-        _pitch_measured_lpf.reset(_ahrs.get_pitch());
+        if (!_reset_after_takeoff) {
+            _flags.reset          = true;
```

Comment:
Note that _flags.reset is also used in the update_50hz function to reset the complementary filter states. Because  update_50hz is called at a different rate and from a different part of the plane code to update_pitch_throttle we need to check if _flags.reset  is cleared from within update_pitch_throttle._initialise_states before the next frame of update_50hz runs

Reply:
So, if I understand correctly, your concern is that `_flags.reset` could be cleared by `update_pitch_throttle()` before `update_50hz()` has a chance to use it?

In principle, `update_50hz()` is called by the `Plane::update_speed_height()` task, scheduled at 50Hz, while `update_pitch_throttle()` is called by the `Plane::update_alt()` task, scheduled at 10Hz.
So unless the scheduler goes terribly wrong, the execution should go something like that:
1. `Plane::update_alt()` calls `update_pitch_throttle()` and sets the `_flags.reset` bit.
2. `Plane::update_speed_height()` runs between 4 to 5 times, each one calling `update_50hz()` with `_flags.reset == true`.
3. Then `Plane::update_alt()` will run once more, invoking `update_pitch_throttle()` and clearing setting `_flags.reset = false`.

Are you satisfied with this argument?

Reply:
@priseborough my explanation above is wrong, because `update_50hz()` always sets the value of `_flags.reset` right at its beginning.

Still, I think what you really meant was that since `update_50hz()` has its own, separate criteria for setting `_flags.reset` and that would affect its use in `update_pitch_throttle()`.
But this is also not the case, because the first thing `update_pitch_throttle() -> _initialize_states()` does is to set `_flags.reset = false`. From then on, it decides whether it needs to set to true for the rest of the loop.

Reply:
Should be OK